src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,181 @@
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly scheduledMethod: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.scheduledMethod.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
+
+            // Init workflow
+            const workflowUUID = `sched-${this.scheduledMethod.name}-${nextExecTime.toISOString()}`;
+            this.dbosExec.logger.debug(`Executing scheduled workflow ${workflowUUID}`);
+            const wfParams = { workflowUUID: workflowUUID };
+            // All operations annotated with Scheduled decorators must take in these four
+            const args: ScheduledArgs = [nextExecTime, new Date(), 0, 0]; // TODO calculate outstanding numbers
+
+            // We can only guarantee exactly-once-per-message execution of transactions and workflows.
+            if (this.scheduledMethod.workflowConfig) {
+                // Execute the transaction
+                await this.dbosExec.workflow(this.scheduledMethod.registeredFunction as Workflow<[Date, Date, number, number], unknown>, wfParams, ...args);
+            }
+            else {
+                this.dbosExec.logger.error(`Function ${this.scheduledMethod.name} is @scheduled but not a workflow`);
+            }
+
+            // Record the time of the wf kicked off
+            const dbTime = await this.dbosExec.systemDatabase.setLastScheduledTime(this.scheduledMethod.name, nextExecTime.getTime());
+            if (dbTime && dbTime > nextExecTime.getTime()) nextExecTime.setTime(dbTime);
```

Comment:
Worth adding a comment here.
Does this mean that another VM is running a future scheduled time so we need to catch up?

Reply:
This means another VM ran a scheduled workflow and we do not need to attempt it.